catalyst_voices/packages/internal/catalyst_voices_services/lib/src/campaign/campaign_service.dart

```diff
@@ -1,13 +1,36 @@
 import 'package:catalyst_voices_models/catalyst_voices_models.dart';
 import 'package:catalyst_voices_repositories/catalyst_voices_repositories.dart';
 
-class CampaignService {
+abstract interface class CampaignService {
+  factory CampaignService(
+    CampaignRepository campaignRepository,
+  ) {
+    return CampaignServiceImpl(
+      campaignRepository,
+    );
+  }
```

Comment:
I'm not sure how this improves code, you can't mock this, you're making the interface know about it's implementation. What are the benefits?

Reply:
- if class has `final` modifier it can not be mocked, `interface` yes. I can just say `class CampaignService` which is mockable but in my opinion its less preferable.
- depending on abstraction rather then concert implementation is always better (SOLID principle)
- from my experience its easier to design class behaviours just by looking at public methods (interface) then at normal class because deves tents to add too many public data

Reply:
I meant the factory constructor which is tied to a specific implementation, not the whole interface. I agree with your points for the interface but this constructor is opposite of what you wrote, isn't it?

Reply:
not exactly. We can have different factory constructor which will return different implementation. 

Main point here was not to expose `CampaignServiceImpl` to other packages. No one know about it beside this file

Reply:
Okay, makes sense